securityscanutils/trivy_scanner.go

```diff
@@ -31,73 +41,71 @@ func RunTrivyScan(image, version, templateFile, output string) (bool, bool, erro
-		// we have since filtered out non-release images from being scanned so this warning
-		// shouldn't occur, but leaving here in case there was another edge case we missed
-		if IsImageNotFoundErr(string(out)) {
-			log.Warnf("image %s not found for version %s", image, version)
-			return false, false, nil
-		}
-		return false, false, eris.Wrapf(err, "error running trivy scan on image %s, version %s, Logs: \n%s", image, version, string(out))
 	}
-	return true, vulnFound, nil
+
+	return scanCompleted, vulnerabilityFound, err
 }
 
-func executeTrivyScanWithRetries(trivyScanArgs []string, retryCount int,
-	backoffStrategy func(int)) ([]byte, int, error) {
-	if retryCount == 0 {
-		retryCount = 5
-	}
-	if backoffStrategy == nil {
-		backoffStrategy = func(attempt int) {
-			time.Sleep(time.Second)
-		}
-	}
+// executeScanWithRetries executes a trivy command (with retries and backoff)
+// and returns a tuple of (scanCompleted, vulnerabilitiesFound, error)
+func (t *TrivyScanner) executeScanWithRetries(ctx context.Context, scanArgs []string) (bool, bool, error) {
+	logger := contextutils.LoggerFrom(ctx)
 
 	var (
 		out        []byte
 		statusCode int
 		err        error
 	)
 
-	for attempt := 0; attempt < retryCount; attempt++ {
-		trivyScanCmd := exec.Command("trivy", trivyScanArgs...)
-		out, statusCode, err = executils.CombinedOutputWithStatus(trivyScanCmd)
+	for attempt := 0; attempt < t.scanMaxRetries; attempt++ {
+		trivyScanCmd := exec.Command("trivy", scanArgs...)
+		attemptStart := time.Now()
+		if t.executeCommand == nil {
+			return false, false, eris.Errorf("TrivyScanner has no executeCommand function defined")
+		}
+		out, statusCode, err = t.executeCommand(trivyScanCmd)
+		logger.Debugf("Trivy returned %d after %s", statusCode, time.Since(attemptStart).String())
 
-		// If there is no error, don't retry
+		// If we receive the expected status code, the scan completed, don't retry
+		if statusCode == VulnerabilityFoundStatusCode {
+			return true, true, nil
+		}
+
+		// If there is no error, the scan completed and no vulnerability was found, don't retry
 		if err == nil {
-			return out, statusCode, err
+			return true, false, err
 		}
 
 		// If there is no image, don't retry
+
 		if IsImageNotFoundErr(string(out)) {
-			return out, statusCode, err
+			logger.Warnf("Trivy scan with args [%v] produced image not found error", scanArgs)
+
+			// swallow error if image is not found error, so that we can continue scanning releases
+			// even if some releases failed and we didn't publish images for those releases
+			// this error used to happen if a release was a pre-release and therefore images
+			// weren't pushed to the container registry.
+			// we have since filtered out non-release images from being scanned so this warning
+			// shouldn't occur, but leaving here in case there was another edge case we missed
+			return false, false, nil
 		}
 
-		backoffStrategy(attempt)
+		t.scanBackoffStrategy(attempt)
```

Comment:
What error could occur that results in a retry? What are we protecting aginst?

Reply:
Added!